relayer/spy_relayer/src/backends/default/listener.ts

```diff
@@ -0,0 +1,272 @@
+  public async parsePayload(
+    rawPayload: Uint8Array
+  ): Promise<ParsedTransferPayload> {
+    let parsedPayload: any;
+    try {
+      parsedPayload = parseTransferPayload(Buffer.from(rawPayload));
+    } catch (e) {
+      this.logger.error("Encountered error while parsing vaa payload" + e);
+    }
+
+    if (!parsedPayload) {
+      this.logger.debug("Failed to parse the transfer payload.");
+      throw new Error("Could not parse the transfer payload.");
+    }
+    return parsedPayload;
+  }
+
+  /** Verify this is a VAA we want to relay. */
+  public async validate(
+    rawVaa: Uint8Array
+  ): Promise<ParsedVaa<ParsedTransferPayload> | string> {
+    let parsedVaa = await this.parseVaa(rawVaa);
+    let parsedPayload: ParsedTransferPayload;
+
+    // Verify this is actual a token bridge transfer payload
+    if (!this.verifyIsPayloadV1(parsedVaa)) {
+      return "Wrong payload type";
+    }
+    try {
+      parsedPayload = await this.parsePayload(parsedVaa.payload);
+    } catch (e: any) {
+      return "Payload parsing failure";
+    }
+
+    // Verify we want to relay this request
+    if (
+      !this.verifyIsApprovedToken(parsedPayload) ||
+      !this.verifyFeeSpecified(parsedPayload)
+    ) {
+      return "Validation failed";
+    }
+
+    // Great success!
+    return { ...parsedVaa, payload: parsedPayload };
+  }
+
+  /** Get spy filters for all emitters we care about */
+  public async getEmitterFilters(): Promise<TypedFilter[]> {
+    let filters: {
+      emitterFilter: { chainId: ChainId; emitterAddress: string };
+    }[] = [];
+    for (let i = 0; i < this.env.spyServiceFilters.length; i++) {
+      this.logger.info("Getting spyServiceFiltera " + i);
+      const filter = this.env.spyServiceFilters[i];
+      this.logger.info(
+        "Getting spyServiceFilter[" +
+          i +
+          "]: chainId = " +
+          filter.chainId +
+          ", emmitterAddress = [" +
+          filter.emitterAddress +
+          "]"
+      );
+      const typedFilter = {
+        emitterFilter: {
+          chainId: filter.chainId as ChainId,
+          emitterAddress: await encodeEmitterAddress(
+            filter.chainId,
+            filter.emitterAddress
+          ),
+        },
+      };
+      this.logger.info("Getting spyServiceFilterc " + i);
+      this.logger.info(
+        "adding filter: chainId: [" +
+          typedFilter.emitterFilter.chainId +
+          "], emitterAddress: [" +
+          typedFilter.emitterFilter.emitterAddress +
+          "]"
+      );
+      this.logger.info("Getting spyServiceFilterd " + i);
+      filters.push(typedFilter);
+      this.logger.info("Getting spyServiceFiltere " + i);
+    }
+    return filters;
+  }
+
+  /** Process and validate incoming VAAs from the spy. */
+  public async process(rawVaa: Uint8Array): Promise<void> {
+    const validationResults: ParsedVaa<ParsedTransferPayload> | string =
+      await this.validate(rawVaa);
```

Comment:
This is one of the lines that I might need to build a type guard for both `ParsedVaa<ParsedTransferPayload>` as I've only got one currently for `ParsedTransferPayload`: https://github.com/certusone/wormhole/blob/3940c95bfb97f7866df92837d6b2f05b9591a24f/relayer/spy_relayer/src/listener/validation.ts#L38-L49